README.md
# Meadow — WordPress templating with Twig

_Write WordPress theme templates with familiar ease and modern features._

[![Code Quality][scrutinizer-ci-badge]][scrutinizer-ci.com]
[![Latest Stable Version][release-badge]][packagist.org]
[![Software License][license-badge]](LICENSE.md)

Meadow is a theme templating solution, aiming to find a balance between native WordPress concepts and power of [Twig](https://twig.symfony.com/) dedicated templating language.

> `mcaskill/wp-meadow` aims to be a drop-in replacement for `rarst/meadow`. It focuses on keeping Twig up to date and enabling any PSR-11 implementation.

## Installation

Require package in your theme project with [Composer](https://getcomposer.org/):

```console
composer require mcaskill/wp-meadow
```

Instantiate object some time during theme load:

```php
$meadow = new \McAskill\Meadow\Core;
$meadow->enable();
```

## Migrating from Rarst

First, replace `rarst/meadow` with `mcaskill/wp-meadow`:

```console
composer remove rarst/meadow
composer require mcaskill/wp-meadow
```

Afterwards, replace the PHP namespace:

```diff
- $meadow = new \Rarst\Meadow\Core;
+ $meadow = new \McAskill\Meadow\Core;
```

At this point your code should mostly work as it did before.

## Templating

Meadow follows conventions of WordPress [template hierarchy](https://codex.wordpress.org/Template_Hierarchy#Visual_Overview):

- for example `index.php` becomes `index.twig`.
- `{{ get_header() }}` will look for `header.twig` (with fallback to `header.php`)
- and so on.

### Template Tags

Template Tags API (and PHP functions in general) are set up to work transparently from Twig templates:

```twig
{{ the_title() }}
```

### Filters

WordPress filters set up to be available as Twig filters:

```twig
{{ 'This is the title' | the_title }}
```

### Template Inheritance

Full range of Twig functionality is naturally available, including [template inheritance](http://twig.sensiolabs.org/doc/templates.html#template-inheritance):

```twig
{# single.twig #}
{% extends 'index.twig' %}

{% block entry_title %}
    <div class="page-header">{{ parent() }}</div>
{% endblock %}
```

To inherit parent template in child theme prepend it with folder's name:

```twig
{# child-theme/index.twig #}
{% extends 'parent-theme/index.twig' %}
```

## Domain Specific Language

Meadow attempts not just "map" WordPress to Twig, but also meaningfully extend both to improve historically clunky WP constructs.

This is primarily achieved by implementing custom Twig tags, abstracting away complexities for specific tasks.

### Loop

```twig
{% loop %}
    <h2><a href="{{ the_permalink() }}">{{ the_title() }}</a></h2>
    {{ the_content() }}
{% endloop %}
```

### Secondary Loop

```twig
{% loop { 'post_type' : 'book', 'orderby' : 'title' } %} {# expression for arguments #}
    <h2><a href="{{ the_permalink() }}">{{ the_title() }}</a></h2>
    {{ the_content() }}
{% endloop %}
```

### Comments

```twig
<ul class="comment-list">
    {% comments %}
        <li>
        {{ comment_text() }}
        {# no </li> - self-closing #}
    {% endcomments %}
</ul>
```

## Template Examples

Check out the [Hybrid Wing](https://github.com/Rarst/hybrid-wing) theme for an example of `rarst/meadow`:

- [`index.twig`](https://github.com/Rarst/hybrid-wing/blob/master/index.twig)
- [`single.twig`](https://github.com/Rarst/hybrid-wing/blob/master/single.twig)
- [`single-post.twig`](https://github.com/Rarst/hybrid-wing/blob/master/single-post.twig)
- [`comments.twig`](https://github.com/Rarst/hybrid-wing/blob/master/comments.twig)

## License

MIT

[scrutinizer-ci-badge]: https://scrutinizer-ci.com/g/mcaskill/wp-meadow/badges/quality-score.png?b=master
[release-badge]:        https://img.shields.io/github/tag/mcaskill/wp-meadow.svg
[license-badge]:        https://poser.pugx.org/mcaskill/wp-meadow/license

[scrutinizer-ci.com]:   https://scrutinizer-ci.com/g/mcaskill/wp-meadow/?branch=master
[packagist.org]:        https://packagist.org/packages/mcaskill/wp-meadow


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/)
and this project adheres to [Semantic Versioning](https://semver.org/).

## Unreleased

### Added

- PHP_CodeSniffer rules to enforce WordPress coding conventions

### Changed

- Renamed namespace from `Rarst` to `McAskill`
- Renamed packge from `rarst/meadow` to `mcaskill/wp-meadow`
- Updated package to reflect replacement
- Formatted codebase according to PHP/WPCS
- Meadow requires WordPress 4.7 or highter

### Removed

- Removed `Template_Hierarchy` class
- Removed `pds/skeleton` as a dev-dependency

## 0.4.0 - TBD

### Added

- Added `.editorconfig` file to enforce coding standards

### Changed

- Updated URL for Twig
- Changed to secure URLs for Keep a Changelog and Semantic Versioning
- Formatted codebase according to EditorConfig
- Cleaned `.gitignore` rules
- Cleaned README and CHANGELOG

## 0.3.0 - 2018-12-31

### Changed

- Code structure for PDS and PSR-4.
- Dependencies (Twig 2+, PHP 7+).

### Fixed

- Possible double output in `template_include`.

## 0.2.0 - 2016-12-10

### Added

- New hierarchy implementation for WP 4.7+

### Deprecated

- Old hierarchy implementation for WP <4.7, to be removed in 1.0

### Fixed

- Deprecated methods in Twig extension

## 0.1.0 - 2015-04-13


src/Core.php
<?php

namespace McAskill\Meadow;

use Pimple\Container;

/**
 * Main plugin class.
 */
class Core extends Container {
    /**
     * @param array $values Optional array of services/options.
     */
    public function __construct( $values = [] ) {
        global $wp_version;

        $defaults                     = [];
        $defaults['twig.options']     = [];
        $defaults['twig.directories'] = [];

        // This needs to be lazy or theme switchers and alike explode it.
        $defaults['twig.loader'] = function ( $meadow ) {
            $stylesheet_dir  = \get_stylesheet_directory();
            $template_dir    = \get_template_directory();
            $calculated_dirs = [
                $stylesheet_dir,
                $template_dir,
                \plugin_dir_path( __DIR__ ) . 'src/twig',
            ];

            // Enables explicit inheritance from parent theme in child.
            if ( $stylesheet_dir !== $template_dir ) {
                $calculated_dirs[] = \dirname( $template_dir );
            }

            $directories = \array_unique(
                \array_merge(
                    $calculated_dirs,
                    $meadow['twig.directories']
                )
            );

            return new \Twig_Loader_Filesystem( $directories );
        };

        $defaults['twig.undefined_function'] = [ $this, 'undefined_function' ];
        $defaults['twig.undefined_filter']   = [ $this, 'undefined_filter' ];

        $defaults['twig.environment'] = function ( $meadow ) {
            $environment      = new \Twig_Environment( $meadow['twig.loader'], $meadow['twig.options'] );
            $meadow_extension = new Extension();
            $environment->addExtension( $meadow_extension );
            $environment->registerUndefinedFunctionCallback( $meadow['twig.undefined_function'] );
            $environment->registerUndefinedFilterCallback( $meadow['twig.undefined_filter'] );

            if ( \defined( 'WP_DEBUG' ) && WP_DEBUG ) {
                $debug_extension = new \Twig_Extension_Debug();
                $environment->addExtension( $debug_extension );
                $environment->enableDebug();
            }

            return $environment;
        };

        $defaults['hierarchy'] = function () {
            return new Type_Template_Hierarchy();
        };

        parent::__construct( \array_merge( $defaults, $values ) );
    }

    /**
     * Handler for undefined functions in Twig to pass them through to PHP and buffer echoing versions.
     *
     * @param string $function_name Name of the function to handle.
     *
     * @return bool|\Twig_Function
     */
    public static function undefined_function( $function_name ) {
        if ( \function_exists( $function_name ) ) {
            return new \Twig_Function(
                $function_name,
                function () use ( $function_name ) {
                    \ob_start();
                    $return = $function_name( ...\func_get_args() );
                    $echo   = \ob_get_clean();

                    return \empty( $echo ) ? $return : $echo;
                },
                [ 'is_safe' => [ 'all' ] ]
            );
        }

        return false;
    }

    /**
     * Handler for fallback to WordPress filters for undefined Twig filters in template.
     *
     * @param string $filter_name Name of the filter to handle.
     *
     * @return bool|\Twig_Filter
     */
    public static function undefined_filter( $filter_name ) {
        return new \Twig_Filter(
            $filter_name,
            function () use ( $filter_name ) {
                return \apply_filters( $filter_name, \func_get_arg( 0 ) );
            },
            [ 'is_safe' => [ 'all' ] ]
        );
    }

    public function enable() {
        /** @var Template_Hierarchy $hierarchy */
        $hierarchy = $this['hierarchy'];
        $hierarchy->enable();
        \add_filter( 'template_include', [ $this, 'template_include' ], 100 );
        \add_filter( 'get_search_form', [ $this, 'get_search_form' ], 9 );
    }

    public function disable() {
        /** @var Template_Hierarchy $hierarchy */
        $hierarchy = $this['hierarchy'];
        $hierarchy->disable();
        \remove_filter( 'template_include', [ $this, 'template_include' ], 100 );
        \remove_filter( 'get_search_form', [ $this, 'get_search_form' ], 9 );
    }

    /**
     * @param string $template Template found by loader.
     *
     * @return string|bool
     */
    public function template_include( $template ) {
        if ( '.twig' === \substr( $template, -5 ) ) {
            /** @var \Twig_Environment $twig */
            $twig = $this['twig.environment'];

            echo $twig->render( \basename( $template ), \apply_filters( 'meadow_context', [] ) );

            exit;
        }

        return $template;
    }

    /**
     * @param string $form Form markup.
     *
     * @return string
     */
    public function get_search_form( $form ) {
        // Because first time it's an action.
        if ( ! \empty( $form ) ) {
            /** @var \Twig_Environment $twig */
            $twig = $this['twig.environment'];

            return $twig->render( 'searchform.twig' );
        }

        return $form;
    }
}


src/Extension.php
<?php

namespace McAskill\Meadow;

/**
 * Meadow extension for Twig with WordPress specific functionality.
 */
class Extension extends \Twig_Extension implements \Twig_Extension_GlobalsInterface {

    public function getFunctions() {
        $options = [
            'needs_environment' => true,
            'needs_context'     => true,
            'is_safe'           => [ 'all' ],
        ];

        $functions = [];

        foreach ( [ 'get_header', 'get_footer', 'get_sidebar', 'get_template_part', 'get_search_form', 'comments_template' ] as $function ) {
            $functions[] = new \Twig_Function( $function, [ $this, $function ], $options );
        }

        return $functions;
    }

    public function getGlobals() {
        global $wp_query;

        return [ $wp_query ];
    }

    public function getTokenParsers() {
        return [
            new Loop_Token_Parser(),
            new Comments_Token_Parser(),
        ];
    }

    public function get_header( \Twig_Environment $env, $context, $name = null ) {
        return $this->get_template( $env, $context, 'header', $name );
    }

    public function get_templates( $slug, $name = null ) {
        $templates = [];

        if ( ! \empty( $name ) ) {
            $templates[] = "{$slug}-{$name}.twig";
        }

        $templates[] = "{$slug}.twig";

        return $templates;
    }

    public function get_footer( \Twig_Environment $env, $context, $name = null ) {
        return $this->get_template( $env, $context, 'footer', $name );
    }

    public function get_sidebar( \Twig_Environment $env, $context, $name = null ) {
        return $this->get_template( $env, $context, 'sidebar', $name );
    }

    public function get_template_part( \Twig_Environment $env, $context, $slug, $name = null ) {
        try {
            $return = \twig_include( $env, $context, $this->get_templates( $slug, $name ) );
            \do_action( "get_template_part_{$slug}", $slug, $name );
        } catch ( \Twig_Error_Loader $e ) {
            \ob_start();
            \get_template_part( $slug, $name );
            $return = \ob_get_clean();
        }

        return $return;
    }

    /**
     * Skips rendering in native function in favor of Plugin->get_search_form() in filter
     *
     * @return string
     */
    public function get_search_form() {
        return \apply_filters( 'get_search_form', true );
    }

    protected function get_template( \Twig_Environment $env, $context, $type, $name = null ) {
        $function_name = "get_{$type}";

        try {
            $return = \twig_include( $env, $context, $this->get_templates( $type, $name ) );
            \do_action( $function_name, $name );
        } catch ( \Twig_Error_Loader $e ) {
            \ob_start();
            $function_name( $name );
            $return = \ob_get_clean();
        }

        return $return;
    }

    public function comments_template( \Twig_Environment $env, $context, $file = 'comments.twig', $separate_comments = false ) {
        try {
            $env->load( $file );
        } catch ( \Twig_Error_Loader $e ) {
            \ob_start();
            \comments_template( '/comments.php', $separate_comments );

            return \ob_get_clean();
        }

        \add_filter( 'comments_template', [ $this, 'return_blank_template' ] );
        \comments_template( '/comments.php', $separate_comments );
        \remove_filter( 'comments_template', [ $this, 'return_blank_template' ] );

        return \twig_include( $env, $context, $file );
    }

    public function return_blank_template() {
        return __DIR__ . '/blank.php';
    }
}


src/Loop_Node.php
<?php

namespace McAskill\Meadow;

use Twig_Node;
use Twig_Compiler;

/**
 * Compiles loop nodes into WP loop with optional query node for secondary loops.
 */
class Loop_Node extends Twig_Node {
    /**
     * @param Twig_Compiler $compiler
     */
    public function compile( Twig_Compiler $compiler ) {
        $compiler->addDebugInfo( $this );

        if ( $this->hasNode( 'query' ) ) {
            $compiler
                    ->write( '$loop = new \WP_Query(' )
                    ->subcompile( $this->getNode( 'query' ) )
                    ->raw( ");\n" )
                    ->write( 'while ( $loop->have_posts() ) : $loop->the_post();' . "\n" ); // TODO nested loops
        } else {
            $compiler->write( 'while ( \have_posts() ) : \the_post();' . "\n" );
        }

        $compiler
                ->subcompile( $this->getNode( 'body' ) )
                ->write( 'endwhile;' . "\n" );
    }
}


src/Loop_Token_Parser.php
<?php

namespace McAskill\Meadow;

use Twig_Token;

class Loop_Token_Parser extends \Twig_TokenParser {
    /**
     * Parses a token and returns a node.
     *
     * @param Twig_Token $token
     *
     * @return Loop_Node
     */
    public function parse( Twig_Token $token ) {
        $nodes  = [];
        $parser = $this->parser;
        $stream = $parser->getStream();

        if ( ! $stream->getCurrent()->test( Twig_Token::BLOCK_END_TYPE ) ) {
            $nodes['query'] = $parser->getExpressionParser()->parseExpression();
        }

        $stream->expect( Twig_Token::BLOCK_END_TYPE );
        $nodes['body'] = $parser->subparse( [ $this, 'decide_loop_end' ], true );
        $stream->expect( Twig_Token::BLOCK_END_TYPE );

        return new Loop_Node( $nodes, [], $token->getLine(), $this->getTag() );
    }

    /**
     * @return string
     */
    public function getTag() {
        return 'loop';
    }

    /**
     * @param Twig_Token $token
     *
     * @return bool
     */
    public function decide_loop_end( Twig_Token $token ) {
        return $token->test( 'endloop' );
    }
}


src/Comments_Token_Parser.php
<?php

namespace McAskill\Meadow;

use Twig_Token;

class Comments_Token_Parser extends \Twig_TokenParser {
    /**
     * Parses a token and returns a node.
     *
     * @param Twig_Token $token
     *
     * @return Comments_Node
     */
    public function parse( Twig_Token $token ) {
        $nodes  = [];
        $parser = $this->parser;
        $stream = $parser->getStream();

        $stream->expect( Twig_Token::BLOCK_END_TYPE );
        $nodes['callback'] = $parser->subparse( [ $this, 'decide_comments_end' ], true );
        $stream->expect( Twig_Token::BLOCK_END_TYPE );

        return new Comments_Node( $nodes, [], $token->getLine(), $this->getTag() );
    }

    /**
     * @return string
     */
    public function getTag() {
        return 'comments';
    }

    /**
     * @param Twig_Token $token
     *
     * @return bool
     */
    public function decide_comments_end( Twig_Token $token ) {
        return $token->test( 'endcomments' );
    }
}


src/Type_Template_Hierarchy.php
<?php

namespace Rarst\Meadow;

/**
 * Prepends template hierarchy with Twig versions of templates.
 */
class Type_Template_Hierarchy {

    /** @var string[] $template_types Template type names to be used for dynamic hooks. */
    public $template_types = array(
        'embed',
        '404',
        'search',
        'taxonomy',
        'frontpage',
        'privacypolicy',
        'home',
        'attachment',
        'single',
        'page',
        'singular',
        'category',
        'tag',
        'author',
        'date',
        'archive',
        'index',
    );

    /** @var string $type Keep track of last processed template type. */
    protected $type = '';

    public function enable() {

        add_filter( 'template_include', array( $this, 'template_include' ), 9 );

        foreach ( $this->template_types as $type ) {
            add_filter( "{$type}_template_hierarchy", array( $this, 'template_hierarchy' ) );
        }
    }

    public function disable() {

        remove_filter( 'template_include', array( $this, 'template_include' ), 9 );

        foreach ( $this->template_types as $type ) {
            remove_filter( "{$type}_template_hierarchy", array( $this, 'template_hierarchy' ) );
        }
    }

    /**
     * @param string[] $templates Array of possible PHP templates, generated by WP core.
     *
     * @return string[] Array of templates, prepended with Twig versions.
     */
    public function template_hierarchy( $templates ) {

        $this->type = substr( current_filter(), 0, - 19 ); // Trim '_template_hierarchy' from end.

        $twig_templates = [];

        foreach ( $templates as $php_template ) {
            if ( '.php' === substr( $php_template, - 4 ) ) {
                $twig_templates[] = substr( $php_template, 0, - 4 ) . '.twig';
            }
        }

        return array_merge( $twig_templates, $templates );
    }

    /**
     * @param string $template Template located by loader after going through hierarchy.
     *
     * @return string
     */
    public function template_include( $template ) {

        return apply_filters( 'meadow_query_template', $template, $this->type );
    }
}


src/Template_Hierarchy.php
<?php

namespace Rarst\Meadow;

/**
 * Augment native template hierarchy with non-PHP template processing.
 *
 * @deprecated 0.2:1.0 Deprecated in favor of better hierarchy filter in WP 4.7+.
 */
class Template_Hierarchy {

    /** @var string[] $template_types Template type names to be used for dynamic hooks. */
    public $template_types = array(
        'embed',
        '404',
        'search',
        'taxonomy',
        'frontpage',
        'privacypolicy',
        'home',
        'attachment',
        'single',
        'page',
        'singular',
        'category',
        'tag',
        'author',
        'date',
        'archive',
        'index',
    );

    protected $mime_type;

    public function enable() {

        add_action( 'template_redirect', array( $this, 'template_redirect' ) );

        foreach ( $this->template_types as $type ) {
            add_filter( "{$type}_template", array( $this, 'query_template' ) );
        }
    }

    public function disable() {

        remove_action( 'template_redirect', array( $this, 'template_redirect' ) );

        foreach ( $this->template_types as $type ) {
            remove_filter( "{$type}_template", array( $this, 'query_template' ) );
        }

        if ( ! empty($this->mime_type) ) {
            remove_filter( "{$this->mime_type[0]}_template", array( $this, 'query_template' ) );
            remove_filter( "{$this->mime_type[1]}_template", array( $this, 'query_template' ) );
            remove_filter( "{$this->mime_type[0]}{$this->mime_type[1]}_template", array( $this, 'query_template' ) );
        }
    }

    public function template_redirect() {

        if ( is_attachment() ) {
            global $posts;

            if ( ! empty( $posts ) && isset( $posts[0]->post_mime_type ) ) {
                $this->mime_type = explode( '/', $posts[0]->post_mime_type );
            }

            add_filter( "{$this->mime_type[0]}_template", array( $this, 'query_template' ) );
            add_filter( "{$this->mime_type[1]}_template", array( $this, 'query_template' ) );
            add_filter( "{$this->mime_type[0]}{$this->mime_type[1]}_template", array( $this, 'query_template' ) );
        }
    }

    /**
     * @param string $fallback
     *
     * @return string
     */
    public function query_template( $fallback ) {

        $type      = substr( current_filter(), 0, - 9 ); // trim '_template' from end
        $templates = array();

        switch ( $type ) {
            case 'embed':

                $object = get_queried_object();

                if ( ! empty( $object->post_type ) ) {

                    $post_format = get_post_format( $object );

                    if ( $post_format ) {
                        $templates[] = "embed-{$object->post_type}-{$post_format}.twig";
                    }

                    $templates[] = "embed-{$object->post_type}.twig";
                }

                $templates[] = 'embed.twig';

                break;

            case 'taxonomy':
                $term = get_queried_object();

                if ( $term ) {
                    $taxonomy    = $term->taxonomy;
                    $templates[] = "taxonomy-{$taxonomy}-{$term->slug}.twig";
                    $templates[] = "taxonomy-{$taxonomy}.twig";
                }

                $templates[] = 'taxonomy.twig';
                break;

            case 'frontpage':
                $templates = array( 'front-page.twig' );
                break;

            case 'home':
                $templates = array( 'home.twig', 'index.twig' );
                break;

            case 'single':
                $object = get_queried_object();

                if ( $object ) {
                    $templates[] = "single-{$object->post_type}.twig";
                }

                $templates[] = 'single.twig';
                break;

            case 'page':
                $page_id  = get_queried_object_id();
//                $template = get_page_template_slug();
                $pagename = get_query_var( 'pagename' );

                if ( ! $pagename && $page_id ) {
                    // If a static page is set as the front page, $pagename will not be set. Retrieve it from the queried object
                    $post     = get_queried_object();
                    $pagename = $post->post_name;
                }

                // TODO page templates
//                if ( $template && 0 === validate_file( $template ) )
//                    $templates[] = $template;

                if ( $pagename ) {
                    $templates[] = "page-{$pagename}.twig";
                }

                if ( $page_id ) {
                    $templates[] = "page-{$page_id}.twig";
                }

                $templates[] = 'page.twig';
                break;

            case 'category':
            case 'tag':
                $term = get_queried_object();

                if ( $term ) {
                    $templates[] = "{$type}-{$term->slug}.twig";
                    $templates[] = "{$type}-{$term->term_id}.twig";
                }

                $templates[] = "{$type}.twig";
                break;

            case 'author':
                $author = get_queried_object();

                if ( $author ) {
                    $templates[] = "author-{$author->user_nicename}.twig";
                    $templates[] = "author-{$author->ID}.twig";
                }

                $templates[] = 'author.twig';
                break;

            case 'archive':
                $post_types = array_filter( (array) get_query_var( 'post_type' ) );

                if ( \count( $post_types ) === 1 ) {
                    $post_type   = reset( $post_types );
                    $templates[] = "archive-{$post_type}.twig";
                }

                $templates[] = 'archive.twig';
                break;

            default:
                $templates = array( "{$type}.twig" );
        }

        $template = $this->locate_template( $templates );

        if ( empty( $template ) ) {
            $template = $fallback;
        }

        return apply_filters( 'meadow_query_template', $template, $type );
    }

    /**
     * Broken out for easier inheritance to customize lookup logic.
     *
     * @param array|string $templates
     *
     * @return string
     */
    public function locate_template( $templates ) {

        return locate_template( $templates );
    }
}
